rust/arrow/src/array/builder.rs

```diff
@@ -236,6 +251,14 @@ impl<T: ArrowPrimitiveType> BufferBuilderTrait<T> for BufferBuilder<T> {
         self.write_bytes(v.to_byte_slice(), 1)
     }
 
+    default fn append_n(&mut self, n: usize, v: T::Native) -> Result<()> {
+        self.reserve(n)?;
+        for _ in 0..n {
+            self.write_bytes(v.to_byte_slice(), 1)?;
+        }
```

Comment:
Any preference to not to use an iterator here?

Reply:
I'm not sure I understand what you mean?

Reply:
I think he means using just a single call to `write_bytes`.

Reply:
That would be incorrect though? The method should write the value `v` `n` times. Calling `write_bytes(v.to_byte_slice(), n)` would write `v` once and then update the array length by n, which would leave things in an odd state.

Reply:
You can use `to_byte_slice` on a `Vec` or slice also.  To be clear, I'm just guessing here.  @vertexclique do you want to expand on your comment or can I merge this?

Reply:
So, `for _ in 0..n` kind of code, even you don't use, checks the boundaries. This is one of the infamous code pieces that Rust has and discourages. In [here]() you can see one of the examples. Range check exists here:
```
mov     rax, qword ptr [rip + core::iter::range::<impl core::iter::traits::iterator::Iterator for core::ops::range::Range<A>>::next@GOTPCREL]
```
This call makes checked_add and which makes range checks. So that is slowing down for big iterations as you saw in the link. Since this is an open issue in the compiler please consider using iterator methods instead of for loops whenever you see.

As you can also see, in this very code 5 cycles spent every range check (in minimal), so that said it is always better to use iterator methods. Bringing the example here :) 
```
(0..n)
.for_each(|_| {
    self.write_bytes(v.to_byte_slice(), 1)?;
})
```

This will be completely pruned in `-C opt-level=3` most probably :)

Reply:
Thanks, I was not aware of this particular compiler bug. In this particular case the codegen at opt-level 2 or higher is identical, I think the problem as described [here](https://github.com/rust-lang/rust/issues/35981) only manifests on nested for loops.

I can update if necessary but error propagation out of the closure if pretty clunky...

Reply:
The place that you are doing byte masking can be a good alternative. Since that place is also accessing with index:
https://github.com/apache/arrow/pull/6980/files#diff-76c0ed0285da0826a55cf62676556c87R102

If that's extreme, just forget about it. :)

Reply:
When unoptimised both forms contain a checked_add, it is just hard to see as the foreach+closure isn't inlined, and the codegen for the iterator form is significantly worse in general. When optimised the codegen is identical for both forms.

Given this, I'm inclined to optimise for readability and use the simpler form, but ultimately it is up to you as the maintainers :)

Reply:
I'm not sure that there is consensus on a single "best" implementation so I'm going move forward with the current implementation as it is more readable.  

@vertexclique feel free to open another JIRA and PR (preferable with a benchmark) if you feel strongly.

p.s. sorry for causing more confusion than anything else above...